src/G2/Equiv/Tactics.hs

```diff
@@ -82,6 +83,315 @@ type Tactic s = s ->
+      l = map (\(i, e) -> (Var i, e)) $ HM.toList hm
+      -- this should only be used with primitive types
+      -- no apparent problems come from using TyUnknown
+      l' = map (\(e1, e2) ->
+                  if (T.isPrimType $ typeOf e1) && (T.isPrimType $ typeOf e2)
+                  then Just $ App (App (Prim Eq TyUnknown) e1) e2
+                  else Nothing) l
+      l'' = [c | Just c <- l']
+      new_conds' = l'' ++ new_conds
+      -- not safe to use unless the lists are non-empty
+      conj_new = foldr1 (\o1 o2 -> App (App (Prim And TyUnknown) o1) o2) new_conds'
+      conj_old = foldr1 (\o1 o2 -> App (App (Prim And TyUnknown) o1) o2) old_conds
+      imp = App (App (Prim Implies TyUnknown) conj_new) conj_old
+      neg_imp = ExtCond (App (Prim Not TyUnknown) imp) True
+      neg_conj = ExtCond (App (Prim Not TyUnknown) conj_old) True
+  res <- if null old_conds
+         then return $ S.UNSAT ()
+         else if null new_conds'
+         then W.liftIO $ LA.applySolver solver (P.insert neg_conj P.empty) s1 s2
+         else W.liftIO $ LA.applySolver solver (P.insert neg_imp P.empty) s1 s2
+  case res of
+    S.UNSAT () -> return True
+    _ -> return False
+
+-- s1 is the old state, s2 is the new state
+-- If any recursively-defined functions or other expressions manage to slip
+-- through the cracks with the other mechanisms in place for avoiding infinite
+-- inlining loops, then we can handle them here by keeping track of all of the
+-- variables that have been inlined previously.
+-- Keeping track of inlinings by adding to ns only lets a variable be inlined
+-- on one side.  We need to have two separate lists of variables that have been
+-- inlined previously so that inlinings on one side do not block any inlinings
+-- that need to happen on the other side.
+-- Whenever a variable is inlined, we record the expression that was on the
+-- opposite side at the time.  Under the original system, a variable could not
+-- be inlined at all on one side in any sub-expressions that resulted from an
+-- inlining of it, and that was too restrictive.  Under the current system,
+-- repeated inlinings of a variable are allowed as long as the expression on
+-- the opposite side is not the same as it was when a previous inlining of the
+-- same variable happened.
+-- When we make the two sides for a new lemma, if the two expressions
+-- contain any variables that aren't present in the expression environment,
+-- we add them to the expression environment as non-total symbolic
+-- variables.  This can happen if an expression for a lemma is a
+-- sub-expression of a Case branch, a Let statement, or a lambda expression
+-- body.  It is possible that we may lose information about the variables
+-- because of these insertions, but this cannot lead to spurious
+-- counterexamples because these insertions apply only to lemmas and lemmas
+-- are not used for counterexample generation.
+moreRestrictive :: StateET
```

Comment:
This function (and a bunch of other functions in G2.Equiv.Tactics) were moved to G2.Language.Approximation.